torch/csrc/jit/passes/tensorexpr_fuser.cpp

```diff
@@ -393,6 +451,131 @@ class TensorExprFuser {
+    for (size_t idx = 0; idx < subgraph->inputs().size(); idx++) {
+      // TODO: Make sure we use types encoded in the IR (not through the
+      // typeinfo map, but explicitly as value types)
+      if (typeinfo_map_.count(subgraph->inputs()[idx])) {
+        subgraph->inputs()[idx]->setType(
+            typeinfo_map_.at(subgraph->inputs()[idx]));
+        inputs_to_check.push_back(n->input(idx));
+      }
+    }
+
+    // Add prim::TypeCheck node
+    Node* typecheck_node =
+        n->owningGraph()
+            ->create(
+                prim::TypeCheck, inputs_to_check, inputs_to_check.size() + 1)
+            ->insertBefore(n);
+    Value* typecheck_result = typecheck_node->output(inputs_to_check.size());
+
+    // Fixup types of the typecheck node outputs
+    std::unordered_map<Value*, Value*> typechecked_inputs;
+    size_t output_idx = 0;
+    for (size_t idx = 0; idx < subgraph->inputs().size(); idx++) {
+      if (typeinfo_map_.count(subgraph->inputs()[idx])) {
+        typechecked_inputs[n->input(idx)] = typecheck_node->output(output_idx);
+        typecheck_node->output(output_idx++)
+            ->setType(typeinfo_map_.at(subgraph->inputs()[idx]));
+      }
+    }
+    typecheck_node->output(inputs_to_check.size())->setType(BoolType::get());
+
+    // Insert if
+    auto versioning_if =
+        n->owningGraph()
+            ->create(prim::If, {typecheck_result}, n->outputs().size())
+            ->insertAfter(typecheck_node);
+    for (size_t idx = 0; idx < n->outputs().size(); ++idx) {
+      n->output(idx)->replaceAllUsesWith(versioning_if->output(idx));
+    }
+    auto true_block = versioning_if->addBlock();
+    auto false_block = versioning_if->addBlock();
+
+    // Fill in the false block. It should contain the unoptimized
+    // copy of the fused subgraph.
+    WithInsertPoint guard(false_block->return_node());
+    const auto subgraph_outputs =
+        insertGraph(*n->owningGraph(), *subgraph, n->inputs());
+    for (Value* output : subgraph_outputs) {
+      false_block->registerOutput(output);
+    }
+
+    // Fill in the true block. It has all inputs type-checked and its
+    // body should be the fusion group node.
+    n->moveBefore(true_block->return_node());
+    for (size_t idx = 0; idx < n->inputs().size(); ++idx) {
+      if (typechecked_inputs.count(n->input(idx))) {
+        n->replaceInput(idx, typechecked_inputs.at(n->input(idx)));
+      }
+    }
+    for (Value* output : n->outputs()) {
+      true_block->registerOutput(output);
+    }
+  }
+
+  void guardFusionGroups(Block* block) {
+    std::vector<Node*> fusion_groups;
+    for (Node* n : block->nodes()) {
+      for (Block* b : n->blocks()) {
+        guardFusionGroups(b);
+      }
+      if (n->kind() == getTensorExprSymbol()) {
+        fusion_groups.push_back(n);
+      }
+    }
+    for (Node* fusion_group : fusion_groups) {
+      guardFusionGroup(fusion_group);
+    }
+  }
+
+  std::unordered_map<Value*, TensorTypePtr> typeinfo_map_;
```

Comment:
maybe add a comment somewhere that the specialized tensors will be Value * in the graph before they are merged, and Value * in the fusion subgraph after they are merged